### PR TITLE
[Notifier] Add GatewayApi bridge

### PR DESCRIPTION
**Pull request by OskarStark**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Replaces #38685
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14463 
| Recipe PR        | https://github.com/symfony/recipes/pull/864

Initial PR by @PGLongo

### CODE REVIEW

**Comment by OskarStark** [comment]
I suppose test are green after https://github.com/symfony/symfony/pull/39571 🤔 

cc @chalasr

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                'message' => $message->getSubject(),
+            ],
+        ]);
+
+        $statusCode = $response->getStatusCode();
+        if (200 !== $statusCode) {
+            throw new TransportException(sprintf('Unable to send the SMS: error %d.', $statusCode), $response);
+        }
+
+        return new SentMessage($message, (string) $this);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]: @PGLongo do we receive a unique identifier in the `$response` which we could set like this ❓ 
```suggestion
        $sentMessage = new SentMessage($message, (string) $this);
        $content = $response->toArray();
        $sentMessage->setMessageId((string) $content['ids'][0]);
        
        return $sentMessage;
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        $statusCode = $response->getStatusCode();
+        if (200 !== $statusCode) {
+            throw new TransportException(sprintf('Unable to send the SMS: error %d.', $statusCode), $response);
+        }
+
+        $sentMessage = new SentMessage($message, (string) $this);
+        $content = $response->toArray();
+        $sentMessage->setMessageId((string) $content['ids'][0]);
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]:
```suggestion

```

**Comment by OskarStark** [comment]
Ready to merge

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        $statusCode = $response->getStatusCode();
+        if (200 !== $statusCode) {
+            throw new TransportException(sprintf('Unable to send the SMS: error %d.', $statusCode), $response);
+        }
+
+        $sentMessage = new SentMessage($message, (string) $this);
+        $content = $response->toArray();
+        $sentMessage->setMessageId((string) $content['ids'][0]);
+        return $sentMessage;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransport.php]:
```suggestion
        $sentMessage->setMessageId((string) $content['ids'][0]);

        return $sentMessage;
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    public function create(Dsn $dsn): TransportInterface
+    {
+        $scheme = $dsn->getScheme();
+
+        if ('gatewayapi' !== $scheme) {
+            throw new UnsupportedSchemeException($dsn, 'gatewayapi', $this->getSupportedSchemes());
+        }
+
+        $authToken = $this->getUser($dsn);
+        $from = $dsn->getOption('from');
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]:
```suggestion
        $from = $dsn->getRequiredOption('from');
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        if (!$from) {
+            throw new IncompleteDsnException('Missing from.', $dsn->getOriginalDsn());
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]:
```suggestion

```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        $this->expectException(IncompleteDsnException::class);
+
+        $factory->create(Dsn::fromString('gatewayapi://host.test?from=Symfony'));
+    }
+
+    public function testCreateWithNoFromShouldThrowsIncompleteDsnException()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(IncompleteDsnException::class);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        $this->expectException(MissingRequiredOptionException::class);
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    public function testCreateWithNoTokenThrowsIncompleteDsnException()
+    {
+        $factory = $this->createFactory();
+
+        $this->expectException(IncompleteDsnException::class);
+
+        $factory->create(Dsn::fromString('gatewayapi://host.test?from=Symfony'));
+    }
+
+    public function testCreateWithNoFromShouldThrowsIncompleteDsnException()
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
    public function testCreateWithNoFromShouldThrowsMissingRequiredOptionException()
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Symfony\Component\Notifier\Bridge\GatewayApi\Tests;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Notifier\Bridge\GatewayApi\GatewayApiTransportFactory;
+use Symfony\Component\Notifier\Exception\IncompleteDsnException;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
use Symfony\Component\Notifier\Exception\IncompleteDsnException;
use Symfony\Component\Notifier\Exception\MissingRequiredOptionException;
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Bridge\GatewayApi;
+
+use Symfony\Component\Notifier\Exception\IncompleteDsnException;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/GatewayApiTransportFactory.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/LICENSE [hunk 10]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2020 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/LICENSE]:
```suggestion
Copyright (c) 2021 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function incompleteDsnProvider(): iterable
+    {
+        yield 'missing token' => ['gatewayapi://host.test?from=Symfony'];
+    }
+
+    public function missingRequiredOptionProvider(): iterable
+    {
+        yield 'missing option: from' => ['gatewayapi://token@host.test?from=Symfony'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        yield 'missing option' => ['gatewayapi://token@host.test'];
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        $transport = $this->createTransport($client);
+        $message = new SmsMessage('3333333333', 'Hello!');
+        $sentMessage = $transport->send($message);
+
+        $this->assertInstanceOf(SentMessage::class, $sentMessage);
+        $this->assertSame('42', $sentMessage->getMessageId());
+    }
+
+
+}
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportTest.php]:
```suggestion
}
```

**Comment by OskarStark** [comment]
Ready to merge 👍

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "~5.3.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/composer.json]:
```suggestion
        "symfony/notifier": "^5.3"
```

**Comment by derrabus** [comment]
@OskarStark Can you have a look at the tests?

**Comment by OskarStark** [comment]
Done @derrabus , lets see if its working now.

Btw, its weird that one passed... 🤔 
CleanShot 2021-01-07 at 11 04 15

And fabbot.io is complaining:
CleanShot 2021-01-07 at 11 05 35

**Comment by OskarStark** [comment]
Tests green 💚

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/CHANGELOG.md [hunk 14]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.3.0
+-----
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/CHANGELOG.md]:
```suggestion
5.3
---
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/CHANGELOG.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+
+ * Added the bridge
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/CHANGELOG.md]:
```suggestion
 * Add the bridge
```

File: src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function incompleteDsnProvider(): iterable
+    {
+        yield 'missing token' => ['gatewayapi://host.test?from=Symfony'];
+    }
+
+    public function missingRequiredOptionProvider(): iterable
+    {
+        yield 'missing option' => ['gatewayapi://token@host.test'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/GatewayApi/Tests/GatewayApiTransportFactoryTest.php]:
```suggestion
        yield 'missing option: from' => ['gatewayapi://token@host.test'];
```

**Comment by OskarStark** [comment]
Travis failure unrelated

**Comment by OskarStark** [comment]
Thank you Piergiuseppe Longo.

**Pull request merged**